### ISSUE TITLE
Bug iterating through TaylorN

### ISSUE THREAD

**Issue by dpsanders** [JuliaDiff/TaylorSeries.jl#140]
Probably to do with the recent changes to indexing. Shouldn't the first HomogeneousPolynomial have order 0?
```jl
julia> using TaylorSeries

julia> a, b = set_variables("a b")
2-element Array{TaylorSeries.TaylorN{Float64},1}:
  1.0 a + 𝒪(‖x‖⁷)
  1.0 b + 𝒪(‖x‖⁷)

julia> a.order
6

julia> for h in a
           @show h, h.order
       end
(h, h.order) = ( 0.0 + 𝒪(‖x‖²), 1)
(h, h.order) = ( 1.0 a + 𝒪(‖x‖³), 2)
(h, h.order) = ( 0.0 + 𝒪(‖x‖⁴), 3)
(h, h.order) = ( 0.0 + 𝒪(‖x‖⁵), 4)
(h, h.order) = ( 0.0 + 𝒪(‖x‖⁶), 5)
(h, h.order) = ( 0.0 + 𝒪(‖x‖⁷), 6)
ERROR: AssertionError: order ≤ get_order() && lencoef ≤ num_coeffs
Stacktrace:
 [1] resize_coeffsHP!(::Array{Float64,1}, ::Int64) at /Users/davidsanders/.julia/v0.6/TaylorSeries/src/auxiliary.jl:37
 [2] Type at /Users/davidsanders/.julia/v0.6/TaylorSeries/src/constructors.jl:96 [inlined]
 [3] Type at /Users/davidsanders/.julia/v0.6/TaylorSeries/src/constructors.jl:102 [inlined]
 [4] macro expansion at /Users/davidsanders/.julia/v0.6/TaylorSeries/src/arithmetic.jl:49 [inlined]
 [5] macro expansion at ./simdloop.jl:73 [inlined]
 [6] zeros(::TaylorSeries.HomogeneousPolynomial{Float64}, ::Int64) at /Users/davidsanders/.julia/v0.6/TaylorSeries/src/arithmetic.jl:48
 [7] TaylorSeries.TaylorN{Float64}(::Array{TaylorSeries.HomogeneousPolynomial{Float64},1}, ::Int64) at /Users/davidsanders/.julia/v0.6/TaylorSeries/src/constructors.jl:157
 [8] next(::TaylorSeries.TaylorN{Float64}, ::Int64) at /Users/davidsanders/.julia/v0.6/TaylorSeries/src/auxiliary.jl:195
 [9] anonymous at ./<missing>:?
```

**Comment by dpsanders** [comment]
The problem seems to be at https://github.com/JuliaDiff/TaylorSeries.jl/blob/5d9794fedf32b9d67a9f44ce1f4a367e6f453df0/src/auxiliary.jl#L196

It should now be `>=` instead of `>`. This will need a test.

**Comment by lbenet** [comment]
Thanks for reporting. Indeed, there is a problem.

I think the problem is related to the line before the one you pointed out:

https://github.com/JuliaDiff/TaylorSeries.jl/blob/5d9794fedf32b9d67a9f44ce1f4a367e6f453df0/src/auxiliary.jl#L195

because the displayed order is incorrect.

**Comment by lbenet** [comment]
I think the line should be

```julia
next(a::$T, ord) = ($T(a[ord], ord), ord+1)
```

which then yields
```julia
julia> for h in a
           @show h, h.order
       end
(h, h.order) = ( 0.0 + 𝒪(‖x‖¹), 0)
(h, h.order) = ( 1.0 a + 𝒪(‖x‖²), 1)
(h, h.order) = ( 0.0 + 𝒪(‖x‖³), 2)
(h, h.order) = ( 0.0 + 𝒪(‖x‖⁴), 3)
(h, h.order) = ( 0.0 + 𝒪(‖x‖⁵), 4)
(h, h.order) = ( 0.0 + 𝒪(‖x‖⁶), 5)
(h, h.order) = ( 0.0 + 𝒪(‖x‖⁷), 6)
```

**Comment by lbenet** [comment]
See #141...

What kind of tests do you have in mind?

**Comment by dpsanders** [comment]
Just checking exactly what you wrote above.
Probably the easiest way is to check the result of `[h for h in a]`.

**Comment by dpsanders** [comment]
And `[h.order for h in a]`.

**Comment by lbenet** [comment]
A silly question: what are we actually after with all this? 

While the following might be correct, I am not sure if that this is what we are actually after...
```julia
julia> a = Taylor1([1,2,3,4,5])
 1 + 2 t + 3 t² + 4 t³ + 5 t⁴ + 𝒪(t⁵)

julia> for h in a
           @show h, h.order
       end
(h, h.order) = ( 1 + 𝒪(t¹), 0)
(h, h.order) = ( 2 + 𝒪(t²), 1)
(h, h.order) = ( 3 + 𝒪(t³), 2)
(h, h.order) = ( 4 + 𝒪(t⁴), 3)
(h, h.order) = ( 5 + 𝒪(t⁵), 4)
```

My point is that `h` above is a `Taylor1` of correct order (so far so good), which has a constant value(!). This is related to `next`:
```julia
julia> next(a, 3)
( 4 + 𝒪(t⁴), 4)
```
Shouldn't we rather get something like `( 4 t³ + 𝒪(t⁴), 4)`

**Comment by lbenet** [comment]
The [documentation](https://docs.julialang.org/en/stable/stdlib/collections/#Base.next) states that `next` should return the *current item* (of the iterable object) and the next state. So, in the example `next(a, 3)`, shouldn't we expect to get simply `4`?

**Comment by lbenet** [comment]
One more observation: `[h.order for h in a]` returns a `0-dimensional Array{Int64,0}`...

**Comment by lbenet** [comment]
I've been playing further with this, and noticed other inconsistencies. For instance, 
```julia
julia> a = Taylor1([1,2,3,4,5])
 1 + 2 t + 3 t² + 4 t³ + 5 t⁴ + 𝒪(t⁵)

julia> for ac in a
           @show ac
       end
ac =  1 + 𝒪(t¹)
ac =  2 + 𝒪(t²)
ac =  3 + 𝒪(t³)
ac =  4 + 𝒪(t⁴)
ac =  5 + 𝒪(t⁵)

julia> Taylor1(1,0) in a # see first entry above
false

julia> 1 in a
false

julia> 1 in a[:]
true
```

I think the inconsistencies are related to the fact that all types (`Taylor1`, etc) are subtypes of `Number`, so some methods (e.g., `in`) use the defaults provided in Base, without noticing that we would like to treat them as iterable.

So, I think #141 should not attempt to fix `next` as proposed, but simply eliminate overloading it, together with `start` and `done`. This will force to use `h in a[:]` instead of `h in a` which attempts to treat `a` as iterable. I think this is more natural.

@dpsanders What do you think?

**Comment by lbenet** [comment]
@dpsanders Do you have any opinion on my last comment?

**Comment by dpsanders** [comment]
Note that the function `in` is not the same as `for x in a`, which is just syntax for iterating using `start`, `next` and `done`. The problem seems to be in `next` etc. defined on `Taylor1`s.

I think it would be a shame to remove this iteration mechanism; we should just make sure it works correctly ;)

**Comment by lbenet** [comment]
I see your point. Yet, for `a = Taylor1([1,2,3,4,5])` it would be nice to have `1 in a` returning true. 

My point with the above comments is simply that by making `Taylor1` a subtype of `Number` we are inheriting some methods; using `start`, `next` and `done` we are defining them as iterables. I think we should make iterable `a[:]` but not `a`.

**Comment by dpsanders** [comment]
Why would you want `a` not to be iterable though?

For `1 in a` we have currently
```
julia> @which 1 in a
in(x::Number, y::Number) in Base at number.jl:121
```
as you suggested.
We just need to write `in` correctly for `Taylor1` if you want `1 in a` to work (though I'm not sure what the use case for that is?)

Although the iteration for `Taylor1` does seem pretty broken right now.

**Comment by lbenet** [comment]
It is not that I don't want it (or do want it) iterable. I think we can simply focus on `a[i0:i1]` or `a[:]` as iterables.

**Comment by dpsanders** [comment]
OK that makes sense.

**Comment by lbenet** [comment]
So, I'll send a PR commenting the implementation of `start`, `next` and `done` later. I think it is worth commenting (instead of simply removing) because of the discussion we had here.

**Comment by lbenet** [comment]
Solved by #141.
